codegen/masm/src/codegen/scheduler.rs

```diff
@@ -0,0 +1,1404 @@
+        loops: &'a LoopAnalysis,
+        liveness: &'a LivenessAnalysis,
+    ) -> Self {
+        Self {
+            f,
+            f_prime,
+            domtree,
+            loops,
+            liveness,
+            schedule: Schedule::new(),
+        }
+    }
+
+    pub fn build(mut self) -> Schedule {
+        self.precompute_block_infos();
+
+        let entry_block_id = self.f.dfg.entry_block();
+        let mut blockq = SmallVec::<[hir::Block; 8]>::from_slice(self.domtree.cfg_postorder());
+        while let Some(block_id) = blockq.pop() {
+            let is_entry_block = block_id == entry_block_id;
+            let schedule = &mut self.schedule.block_schedules[block_id];
+            if is_entry_block {
+                schedule.push(ScheduleOp::Init(block_id));
+            } else {
+                schedule.push(ScheduleOp::Enter(block_id));
+            }
+
+            let block_info = self.schedule.block_infos.get(block_id).cloned().unwrap();
+            let block_scheduler = BlockScheduler {
+                f: self.f,
+                liveness: self.liveness,
+                schedule,
+                block_info,
+                inst_infos: Default::default(),
+                worklist: SmallVec::from_iter([Plan::Start(block_id)]),
+            };
+            block_scheduler.schedule();
+        }
+
+        self.schedule
+    }
+
+    fn precompute_block_infos(&mut self) {
+        let entry_block_id = self.f.dfg.entry_block();
+
+        for block_id in self.domtree.cfg_postorder().iter().rev().copied() {
+            // Ensure we have a target block for each source IR block being scheduled
+            let masm_block_id = if block_id == entry_block_id {
+                self.f_prime.body.id()
+            } else {
+                self.f_prime.create_block()
+            };
+
+            // Set the controlling loop
+            let loop_header = self.loops.is_loop_header(block_id);
+            let last_inst = self.f.dfg.last_inst(block_id).unwrap();
+            let innermost_loop = self.loops.innermost_loop(block_id);
+            let depgraph = build_dependency_graph(block_id, self.f, self.liveness);
+            let treegraph = OrderedTreeGraph::new(&depgraph)
+                .expect("unable to topologically sort treegraph for block");
+
+            let info = Rc::new(BlockInfo {
+                source: block_id,
+                target: masm_block_id,
+                last_inst,
+                innermost_loop,
+                loop_header,
+                depgraph,
+                treegraph,
+            });
+
+            self.schedule.block_infos.insert(info);
+        }
+    }
+}
+
+struct BlockScheduler<'a> {
+    f: &'a hir::Function,
+    liveness: &'a LivenessAnalysis,
+    schedule: &'a mut Vec<ScheduleOp>,
+    block_info: Rc<BlockInfo>,
+    inst_infos: SparseMap<hir::Inst, Rc<InstInfo>>,
+    worklist: SmallVec<[Plan; 4]>,
+}
+impl<'a> BlockScheduler<'a> {
+    pub fn schedule(mut self) {
```

Comment:
I prefer to keep this `pub` to indicate that it is the "public" interface of `BlockScheduler`, even though it is never called outside this module. Otherwise it gets a bit murky as to which functions are meant to be called from outside `BlockScheduler` (IMO).